spices/SPICE-0010-overhauled-mapping-listing-typechecks.adoc

```diff
@@ -0,0 +1,432 @@
+
+In order to typecheck these values correctly, unions types of two or more of the same parameterized type must be checked eagerly.
+The following types are eagerly checked:
+
+[source,pkl]
+----
+value1: Listing<Int>|Listing<String>
+value2: Listing<Int>|Listing<String>?
+value3: Listing<Int>|(Listing<String>)
+value4: Mapping<String, Int>|Mapping<String, String>
+value5: List<Listing<String>>|List<Listing<Int>>
+----
+
+The following types retain laziness during typecheck:
+
+[source,pkl]
+----
+value1: Listing<Int>|Int
+value2: Listing<Int>|Mapping<String, String>
+value3: List<Listing<String>>|Set<String> // <1>
+----
+<1> `List` is _shallow_ checked; the `Listing` object itself is evaluated, but its members are not.
+
+=== Optimizations: `unknown`, `Any`, un-parameterized types
+
+Mappings/Listings with type parameter `unknown` or `Any` do not need to have their members casted.
+Therefore, the type cast of a value to `Listing<unknown>` can return the same in-memory object rather than return a xref:_surrogate_objects[surrogate].
+
+[source,pkl]
+----
+myListing: Listing<unknown> = someListing // <1>
+----
+<1> `listing` and `someListing` are the same in-memory object
+
+Un-parameterized `Mapping` and `Listing` objects also do not need to return surrogate objects. `Listing` is sugar for `Listing<unknown>`.
+
+== Compatibility
+
+=== Change in throw behavior
+
+This is a marginally breaking change in the language.
+The following program currently throws, and will not throw after this change is implemented:
+
+[source,pkl]
+----
+local people: Listing<Person> = new {
+  throw("Uh oh")
+}
+
+len = people.length
+----
+
+On the flip-side, the following program does not throw, and will throw after this change is implemented:
+
+[source,pkl]
+----
+people = new Listing<Person> {
+  1
+}
+----
+
+However, this is considered a bug, and throwing here is simply a fix for the bug.
+
+== Future directions
+
+=== Type-casted `Function` types
+
+Another language bug is in typechecks of anonymous functions (lambdas).
+
+This snippet does not throw, but should:
+
+[source,pkl]
+----
+local myFunc: (Int) -> Int = (it) -> it
+result = myFunc.apply("hello")
+----
+
+The fix here should be similar to the fix applied to Listings and Mappings; to cast the provided function into `(Int) -> Int` by returning a _new_ function.
+
+=== Throw when type testing `Function` types
+
+`Function` types can be _casted_, but cannot be _tested_.
+
+[source,pkl]
+----
+myFunc = (it) -> it as (Int) -> Int // <1>
+
+isMyFunc = myFunc is (Int) -> Int // <2>
+----
+<1> Fine; we can wrap the provided function with a new function that asserts the types of the parameter and return value
+<2> This is impossible to check, because Pkl types do not have the concept of subtypes, and runtime functions do not retain their type signatures.
+
+To be sound, this operation should throw when evaluated.
+
+== Alternatives considered
+
+N/A
```

Comment:
Have you considered to preserve the element type of `Listing<X>` when the listing is amended? Although a breaking change, this doesn't break the common case `xs: Listing<X>` where  `xs` is repeatedly amended (because the property type stays `Listing<X>`). And it's much more space/time efficient because it doesn't require adding another surrogate after every amend.

Reply:
We've thought about this in the past, but this would be quite limiting on the usage side.

For example:

```pkl
local dogs: Listing<Dog> = new {
  new { breed = "Greyhound" }
}

animals: Listing<Animal> = (dogs) {
  new Horse { ... } // fail; can only add `Dog` here
}
```

Reply:
I wonder how often this occurs in practice, and if something like the following would be acceptable in such cases:

```
animals: Listing<Animal> = new {
  ...dogs
  new Horse { ... }
}
```

"obj-surrogate-obj-surrogate" is quite an expensive (and sad) data structure for the common case.

Reply:
FYI: I'm switching terminology to "delegate" for this, because "surrogate" turned out to be kind of a confusing term when discussing this concept internally. Will update docs.

I don't think this is quite so bad for perf. You only get one delegate per typecast, and in the normal case, this happens infrequently.

Idiomatic Pkl code typically looks like this, where there is a schema, and then data that adheres to the schema:

```pkl
// MyPeople
people: Listing<Person>
```

```pkl
// myData.pkl
amends "MySchema.pkl"

people {
  new { name = "Barney" }
}
```

This results in only one delegate when `myData` is eval'd; when `people` is accessed and cast to `Listing<Person>` [^1]. Creating new members within `people` here isn't any more expensive than it is today.

There are two cases that I can think of where this might end up being problematic:

1. There are many collections, rather than many members within a collection. E.g. data that ends up looking like `[[1], [2], [3], [4], [5]]`.
2. A type-cast is executed, where the member type is different, and both the original and the casted object are forced (we end up with two objects, both with their own cached members):
    ```pkl
    people1: Listing<People> = new { ... }
    people2: Listing<People(lastName == "Smith")> = people1
    ```

In any case, I think this is an optimization problem, rather than a problem with the design of the language. Changing this would be a big breaking change, and it would also differ from how this works in most languages (`concat`/`plus` usually accepts contravariant members).

[^1]: Note: `people { ... }` desugars to `people = (super.people) { ... }`, except we do not type-check when accessing `super.people`.

Reply:
Design and implementation should inform each other.

Throwing away all reified type information on every amend is unfortunate from an implementation perspective:
* It is expensive in the common case where a listing is amended and assigned to a property of the same type.
   (Each assignment requires a new delegate.)
* It rules out many potential optimizations, such as using a special representation for `Listing<Int>`.

The current design also means that bugs can go undetected in programs that aren't fully typed.

Fortunately, accepting this SPICE doesn't rule out giving alternative designs a fair shot later on (before 1.0).

Reply:
> It is expensive in the common case where a listing is amended and assigned to a property of the same type.
(Each assignment requires a new delegate.)

Can you elaborate here? It would be helpful to see a code snippet that demonstrates this.

> It rules out many potential optimizations, such as using a special representation for Listing<Int>.

I'm not so sure; imagine we have a special representation of `Listing<Int>` called `VmIntListing`:

```java
class ListingAmendNode {
  @Specialization
  VmIntListing eval(VmIntListing parent, long[] members) {
    // we can keep the specialized representation 
  }

  @Fallback
  VmListing eval(IntListing parent, Object[] members) {
    // need to drop back to generic `VmListing` now
  }
}
```

> Fortunately, accepting this SPICE doesn't rule out giving alternative designs a fair shot later on (before 1.0).

Yup! And this is a orthogonal to this SPICE anyway (how typechecking works vs. how amend is typed).

Reply:
> Can you elaborate here? It would be helpful to see a code snippet that demonstrates this.

```
a: Listing<Int> = new { 0 }
b: Listing<Int> = (a) { 1 }
c: Listing<Int> = (b) { 2 }
```
In the common case that an amended listing is assigned back to a property of the same type (or even the same property), one delegate is created for every amend, doubling the number of `VmListing` objects.

> I'm not so sure; imagine we have a special representation of Listing<Int> called VmIntListing:

Right, that's exactly what I imagined. My point is that an optimized `VmIntListing` representation will very likely want the entire prototype chain to consist of `VmIntListing`.